### ISSUE TITLE
RedisCache Service

### ISSUE THREAD

**Issue by nevenc-pivotal** [Azure/meta-azure-service-broker#166]
I discovered a possible problem with Azure Service Broker (the one that gets created out of the box with Azure Marketplace PCF install):

default RedisCache config is:

```
{
  "parameters": {
      "enableNonSslPort": false,
      "sku": {
          "name": "Basic",
          "family": "C",
          "capacity": 0
      }
  }
}
```

So, non-SSL port is not enabled by default (6379) - and service broker adds both ports to ENV when service is bound, e.g.

```
{
 "VCAP_SERVICES": {
  "azure-rediscache": [
   {
    "credentials": {
     "hostname": "u50060acf075.redis.cache.windows.net",
     "name": "u50060acf075",
     "port": 6379,
     "primaryKey": "xxxx",
     "redisUrl": "redis://u50060acf075:xxxx@u50060acf075.redis.cache.windows.net:6379",
     "secondaryKey": "yyyy",
     "sslPort": 6380
    },
    "label": "azure-rediscache",
    "name": "myredis",
    "plan": "basic",
    "provider": null,
    "syslog_drain_url": null,
    "tags": [
     "Azure",
     "Redis",
     "Cache"
    ],
    "volume_mounts": []
   }
  ]
 }
}
```

Problem is that if nonSSL port is disabled (the case of default Azure Marketplace PCF install):
-  `redisUrl` is using a wrong port (e.g. `port` and `6379` instead of `sslPort` and `6380`)
-   and wrong protocol (e.g. `redis://` instead of `rediss://`)

Things work as expected, if we don't enforce SSL. Service broker creates proper `port` (e.g. `6379`) and `redisUrl` on non-SSL port  (e.g. `redis://....:6379`).

**Comment by zhongyi-zhang** [comment]
Hi @nevenc-pivotal, thanks for raising it.
The `redisUrl` was added to be compatible with Spring Cloud Redis Connector. Up to now, the Connector can only recognize `redis` scheme and doesn't support SSL. (FYI: https://github.com/spring-cloud/spring-cloud-connectors/blob/master/spring-cloud-core/src/main/java/org/springframework/cloud/service/common/RedisServiceInfo.java#L14) So we just deliver `redis://....:6379` and assume that `enableNonSslPort`  would be set to `true` if the user uses Sping Cloud.
But yes, it is still a bug when `enableNonSslPort` is set to `false`.

**Comment by gumichel** [comment]
Hi @zhongyi-zhang,
We are facing this issue in our project. It prevents us from using Redis with Spring Cloud Redis Connector with SSL enable.

**Comment by zhongyi-zhang** [comment]
Hi @gumichel, SSL is always enabled in Azure Redis Cache Service. The option in provisioning parameters is for **non**-SSL port. To use Spring Cloud with Azure Redis, now the only choice is to enable non-SSL because Spring Cloud Redis Connector doesn't support connecting to Redis host with SSL.

**Comment by scottfrederick** [comment]
@zhongyi-zhang Spring Cloud Connectors is pretty flexible in terms of what it can look for in the `credentials` payload for a bound service. It can use discrete host, port, username, and password fields, or it can use a URL. See [RedisServiceInfoCreator](https://github.com/spring-cloud/spring-cloud-connectors/blob/9e1277c7c20e09601305dc4f05e095daf54e566c/spring-cloud-cloudfoundry-connector/src/main/java/org/springframework/cloud/cloudfoundry/RedisServiceInfoCreator.java#L22-L33).

If there are multiple ports (e.g. `port` and `sslPort`) or multiple URLs (e.g. with `redis:` and `rediss:` schemes) in the credentials, there must be some clue in the credentials for Connectors to know which bits of information to use. 

Would it be feasible for the operator to decide at tile installation time, or the developer decide at service instance creation time, whether secure or non-secure connections should be used and provide a single port or URL in the credentials?

**Comment by scottfrederick** [comment]
Also, is the use of `6380` for a secure TLS port a convention followed by other Redis providers, or is that just something that Azure Redis does?

**Comment by zhongyi-zhang** [comment]
> Spring Cloud Connectors is pretty flexible in terms of what it can look for in the credentials payload for a bound service. It can use discrete host, port, username, and password fields, or it can use a URL. See RedisServiceInfoCreator.

@scottfrederick yeah, I got this. The broker always tries to be compatible with Spring Cloud Connectors. Thanks for the explanation.

>If there are multiple ports (e.g. port and sslPort) or multiple URLs (e.g. with redis: and rediss: schemes) in the credentials, there must be some clue in the credentials for Connectors to know which bits of information to use.
Would it be feasible for the operator to decide at tile installation time, or the developer decide at service instance creation time, whether secure or non-secure connections should be used and provide a single port or URL in the credentials?

I think the broker should keep only one URL, the securer one with SSL, in the credentials. Is there a reason to still use non-SSL URL if both the service and the connector support SSL? I'd like to make this change if the connector supports it at upstream.

>Also, is the use of 6380 for a secure TLS port a convention followed by other Redis providers, or is that just something that Azure Redis does?

Looks it is just something Azure Redis does as I commented in https://github.com/spring-cloud/spring-cloud-connectors/issues/235.

**Comment by jmichelgarcia** [comment]
@zhongyi-zhang https://github.com/spring-cloud/spring-cloud-connectors/issues/235 has been fixed and merge into master is in progress. https://github.com/Azure/open-service-broker-azure/pull/564

**Comment by scottfrederick** [comment]
This should be resolved in Spring Cloud Connectors [2.0.3.RELEASE](https://github.com/spring-cloud/spring-cloud-connectors/releases/tag/v2.0.3.RELEASE).

**Comment by zhongyi-zhang** [comment]
@jmichelgarcia @scottfrederick thanks for reminding! I'll fix it with `rediss` schema.

**Comment by zhongyi-zhang** [comment]
Fixed via https://github.com/Azure/meta-azure-service-broker/pull/188.
I've tried `"enableNonSslPort": false` successfully with Spring Music by upgrading the Connector to 2.0.3.RELEASE.

BTW, my Spring Music is at this commit: https://github.com/cloudfoundry-samples/spring-music/commit/71673e61c4118c2d44e6db047f95f31ecf24ae82. I don't know why newer Spring Music with Spring Boot >= 2.0 can't start even when non-SSL port is enabled. The same things work for the history version.
